contributors/design-proposals/secure_watch.md

```diff
@@ -0,0 +1,208 @@
+
+ The crucial requirements here are:
+
+1. Whenever "list" request returns a list of objects and a resource version "rv",
+starting a watch from the returned "rv" will never drop any events.
+2. For a given watch request (with resource version "rv"), the returned stream
+of events is always the same (e.g. very slow laggin watch may not cause dropped
+events). 
+
+ To satisfy the above requirements, we can't really rely only on the existing
+tools/frameworks that we have. Theoretically, we would be able to build the
+mapping from an object to list of nodes it is bound to using standard
+reflector/informer framework. But having separate informers for different kinds
+of objects (which is the only possible way to use them) may result in races.
+As an example, imagine a pod P referencing secret S, bounding pod P to node N
+happening right before creating secret S. Then the following race may potentially
+happen:
+1. user observes binding of pod P, he tries to retrieve the current value of a
+secret and it still doesn't exist. From now she knows that as soon as it will
+be created, it should be delivered via watch to him.
+2. however, pod informer is lagging in our code, and we first observe secret
+creation. It's not referenced yet (we didn't yet observe the pod), so we don't
+send any event.
+3. we observe the pod binding, but according to our semantics, we don't send
+any event.
+And we didn't send the expected "add secret" event to the watcher.
+
+ To solve the problem reliably, we need to be able to correctly serialize the
+watch events between different object types in a deterministic way.
+
+ One potential solution would be to identify this watch by a resource version
+combined from resource versions coming from different object kinds (e.g.
+pods have rv = rv1, secrets have rv = rv2, ...). Then we could keep the history
+of objects being processed to update the in-memory mapping and that would be
+deterministic. But the order might have been different e.g. after restarting
+apiserver. Which also means that it wouldn't work in HA setups (assuming we
+don't have some external "serializing" component).
+
+ Fortunately, we can solve it in much simpler way, with one additional assumption:
+
+1. All object types necessary to determine the in-memory mapping share the same
+resource version series.
+
+ With that assumption, we can have a "multi-object-type" watch that will serialize
+events for different object types for us. Having exactly watch responsible for
+delivering all objects (pods, secrets, ...) will guarantee that if we are currently
+at resource version rv, we procesed object of all types up to rv and nothing with
+resource version greater than rv. Which is exactly what we need.
+
+NOTE: We are not going to implement a generic "multi-object-type" watch and
+expose it in the api (which is much bigger task). This will be purely implementation
+detail hidden in the code (see more details in the next section).
+
+
+### Implementation details
```

Comment:
I haven't yet thought about whether I think this works.  The "pod references PVC which references PV which references secret" and I update the resource links worries me for ordering concerns.  I *think* that in such a case, you have to keep getting the secret until the watch finally returns it you.  Even so, a later update to a PV could remove the secret (you should get a DELETE I think) and a subsequent update to the PV would expose it to you again and we wouldn't be able to show you the ADD because it could be moving back in time.